### DISCUSSION TITLE
Why is find_matches_to_new_records taking so long to run in Spark?

### DISCUSSION THREAD

**Discussion by richard-a-lott** [moj-analytical-services/splink#1444] [Q&A] [answered]
I'm attempting to use 'find_matches_to_new_records' to do some testing of my model (via Glue/Spark), and I'm finding that the jobs are running for around 10 minutes, which is twice as long as the linker.predict() run takes! What could be the cause of this?

My source data set has ~10M records, and my "new records" is just a single entry. It was my expectation that this should run very quickly. In fact a test I've done locally in duckdb on the same data & model takes maybe 2 minutes.

I know that's not much to go on, but I'm not able to share much detail about my data, unfortunately. Any pointers would be helpful though.

**Comment by richard-a-lott** [comment]
Ah so it goes _much_ quicker if I use blocking rules in the new records function, however that presumably means I risk missing potential matches. So I guess I have to find a balance between acceptable speed and finding all matches

**Reply by RobinL** [reply, marked as answer]
Yeas, so by default `find_matches_to_new_records` compares each new record to _all_ records.  So for each record you input, it will make 10M comparisons.

If you're Splink model is fairly simple (e.g. mostly exact matches and few term frequency adjustments), that should be fairly quick.  A more complex model (e.g. with lots of jaro winklers, tf adjustments etc.) will be fairly slow for 10M (e.g. 2 minutes as you say).

As you observe, you can dramatically reduce this 2 minutes by blocking, even with very loose blocking rules. 

e.g. if you were using person data:
```
rules = ["l.first_name = r.first_name",
"l.surname = r.surname",
"l.dob = r.dob"]
```
and it would be very unlikely a truly matching record would 'fall through' all of those criteria (it'd have to be a truly matching record where neither first name, surname, or date of birth matched.

**Comment by richard-a-lott** [comment]
I know I marked this as answered, however I'm still struggling a bit with getting timely predictions in Glue/Spark.

If I do a 'find_matches_to_new_records' prediction in DuckDB on a random 20 rows from my source, it's taking almost 4 minutes. Glue is taking 20 minutes, for 20 rows, with the exact same data, model & blocking rules. So DuckDB is managing about 1 prediction every 12 seconds, compared to one minute for Spark!

Note that linker.predict() is much faster at about 5 minutes.

I'm trying to figure out why that may be. My goal is to run a prediction on say 200 records, but for it to be quicker than my estimated 3 hours! I'd use DuckDB but it's local and not entirely stable with the amount of data in it, hence why I'm looking at Glue.

My Glue job was only allocating 5 executors, so I tried increasing that to 20. Instead of taking 20m it took 14m, not so much of an improvement.

I've tried break linkage method = persist, and I've tried repartition after blocking = True. Neither seem to have any effect.

Any ideas how I can get better performance from Spark? Is it something to do with how the find_matches_to_new_records method works?

**Reply by RobinL** [reply]
I'm try and answer more fully soon, but one thing I think you might find is that increasing to 200 input records doesn't increase the spark runtime very much i.e. it's very nonlinear for small numbers of records.  Might be wrong, it's just a hunch.

**Reply by richard-a-lott** [reply]
Ok that makes sense, thanks. I'll give it a try.

**Reply by RobinL** [reply]
Actually a few more questions that will help me answer better:

- are you using term frequency adjustments in your settings?
- if so, in spark, have you precomputed the tf tables and are loading then in?  If not, I think Splink might recompute them first, which could account for a lot of the runtime.  Sorry the docs on this aren't great but I can provide an example if you need it
- are you now using blocking rules with your find new matches code with the spark script?



One thing you could to try and diagnose is running find new matches twice in your spark script (on different new records) and seeing whether the second call run much faster than the first.

**Reply by richard-a-lott** [reply]
Here you go:
- No I'm not using term frequency adjustments
- Yes I'm using blocking rules, the same for both duckdb and glue. The rules I'm using allow duckdb to produce 1 find new match in about 10 seconds.

Ok I can try that out. I'll also give the 200 records a try to see how long that takes.

**Reply by RobinL** [reply]
Thanks.

Another thing worth trying is setting this a lot lower, to say 10:
https://github.com/moj-analytical-services/splink/blob/31d4d0ec9e5c7f27abc25e57968997841a8d004c/splink/spark/linker.py#L94

And it might be worth setting parallelism lower
```python

# Set default parallelism
spark.conf.set("spark.default.parallelism", "20")

# Set shuffle partitions
spark.conf.set("spark.sql.shuffle.partitions", "20")
```

The defaults for these settings are fairly well optimized for large linking jobs but possibly not very well optimized for find new matches

**Reply by RobinL** [reply]
You could also try setting this to "persist":
https://github.com/moj-analytical-services/splink/blob/31d4d0ec9e5c7f27abc25e57968997841a8d004c/splink/spark/linker.py#L86

**Reply by richard-a-lott** [reply]
Thanks for your help with this @RobinL. After experimenting with your above tweaks, and not seeing much improvement, I've revisited the blocking rules I'm using and this seems to have made the difference.

Cumulatively my blocking rules were generating over a billion comparisons for the entire data set, which is clearly too many. It's interesting though that DuckDB was able to relatively quickly generate predictions, and Spark was not. Any idea why this may be?

However I've trimmed the blocking rules down and now I have Spark able to produce the required 200 predictions in a couple of minutes.

**Reply by RobinL** [reply]
Hmm - interesting.  Sorry nothing much has helped so far!

Spark in general is more suited for large bulk operations.  There's an overhead involved in splitting up the workload across multiple computers, and then having to transfer (shuffle) data between executors for different parts of the computation.   I guess for that reason we envisaged the find new matches function was probably best suited to duckdb; but I would have hoped for better Spark performance.

But I'm a little surprised you're not seeing much change with different settings.

Are you loading the input data from a single large file (perhaps a csv)?  Is it possible that the data load itself takes a long time?  If the input data is a single file, the read can't easily be parallelised, whereas if it's split into several files (generally aim for about 100Mb/file), the read can be parallelised.

**Reply by richard-a-lott** [reply]
I didn't mean to suggest you've not been helpful! It's helped me realise that the issue is probably more to do with me not doing things in the most efficient way. ;)

I have got things running fast enough (3 minutes) by tightening my blocking rules, and I'm happy I'm not losing any matches this way as well.

For reference I'm loading parquet files out of S3, there's around 20 files @ 30MB each. Maybe I can tighten this up a little, but like I say things are running well enough for me now.

**Reply by RobinL** [reply]
No - that's fine!  It's more that it's just piqued my interest - getting this sort of feedback is invaluable because often others are using parts of the codebase that we haven't performance tested thoroughly.  It sounds like your setup is fine, and I'm glad it's working well enough for you now.  

I've [made a note](https://github.com/moj-analytical-services/splink/issues/1479) to have a more detailed look at some point to see if there's anything we can do on the Spark side to make it run faster.
